pkg/util/docker/dockerfile/builder/client.go

```diff
@@ -0,0 +1,388 @@
+	if e.Container == nil {
+		return nil
+	}
+	err := e.Client.RemoveContainer(docker.RemoveContainerOptions{
+		ID:            e.Container.ID,
+		RemoveVolumes: true,
+		Force:         true,
+	})
+	if _, ok := err.(*docker.NoSuchContainer); err != nil && !ok {
+		return err
+	}
+	e.Container = nil
+	return nil
+}
+
+// CreateScratchImage creates a new, zero byte layer that is identical to "scratch"
+// except that the resulting image will have two layers.
+func (e *ClientExecutor) CreateScratchImage() (string, error) {
+	random := make([]byte, 16)
+	if _, err := io.ReadFull(rand.Reader, random); err != nil {
+		return "", err
+	}
+	name := fmt.Sprintf("scratch-%s", base64.URLEncoding.EncodeToString(random))
+
+	buf := &bytes.Buffer{}
+	w := tar.NewWriter(buf)
+	w.Close()
+
+	return name, e.Client.ImportImage(docker.ImportImageOptions{
+		Repository:  name,
+		Source:      "-",
+		InputStream: buf,
+	})
+}
+
+// CleanupImage attempts to remove the provided image.
+func (e *ClientExecutor) CleanupImage(name string) error {
+	return e.Client.RemoveImage(name)
+}
+
+// LoadImage checks the client for an image matching from. If not found,
+// attempts to pull the image and then tries to inspect again.
+func (e *ClientExecutor) LoadImage(from string) (*docker.Image, error) {
+	image, err := e.Client.InspectImage(from)
+	if err == nil {
+		return image, nil
+	}
+	if err != docker.ErrNoSuchImage {
+		return nil, err
+	}
+
+	var registry string
+	repository, tag := docker.ParseRepositoryTag(from)
+	if parts := strings.SplitN(repository, "/", 2); len(parts) > 1 {
+		registry, repository = parts[0], parts[1]
+	}
+
+	// TODO: we may want to abstract looping over multiple credentials
+	auth, _ := e.AuthFn(repository)
+	if len(auth) == 0 {
+		auth = append(auth, docker.AuthConfiguration{})
+	}
+
+	var lastErr error
+	for _, config := range auth {
+		// TODO: handle IDs?
+		err = e.Client.PullImage(docker.PullImageOptions{
+			Registry:   registry,
+			Repository: repository,
+			Tag:        tag,
+		}, config)
+		if err == nil {
+			break
+		}
+		lastErr = err
+		continue
+	}
+	if lastErr != nil {
+		return nil, lastErr
+	}
+
+	return e.Client.InspectImage(from)
+}
+
+// Run executes a single Run command against the current container using exec().
+// Since exec does not allow ENV or WORKINGDIR to be set, we force the execution of
+// the user command into a shell and perform those operations before. Since RUN
+// requires /bin/sh, we can use both 'cd' and 'export'.
+func (e *ClientExecutor) Run(run Run, config docker.Config) error {
+	args := make([]string, len(run.Args))
+	copy(args, run.Args)
+
+	if runtime.GOOS == "windows" {
+		if len(config.WorkingDir) > 0 {
+			args[0] = fmt.Sprintf("cd %s && %s", bashQuote(config.WorkingDir), args[0])
+		}
+		// TODO: implement windows ENV
+		args = append([]string{"cmd", "/S", "/C"}, args...)
+	} else {
+		if len(config.WorkingDir) > 0 {
+			args[0] = fmt.Sprintf("cd %s && %s", bashQuote(config.WorkingDir), args[0])
```

Comment:
I don't think so, it won't make the code cleaner and will contain spurious output.  Can you think of why cd to an absolute path wouldn't work?